File: compiler/nodes/patterns/annotate.go
package patterns

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type AnnotatePatternNode struct {
	Pattern database.Node
	Type    database.Node
	Facts   *database.Facts

	matching database.Node
}

func (node *AnnotatePatternNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseAnnotatePattern(parser *syntax.Parser) (*AnnotatePatternNode, *syntax.Error) {
	span := parser.Spanned()

	left, err := ParsePatternElement(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("AnnotateOperator", syntax.TokenConfig{
		Commit: "in this type annotation",
	})
	if err != nil {
		return nil, err
	}

	parser.ConsumeLineBreaks()

	right, err := types.ParseTypeElement(parser)
	if err != nil {
		return nil, err
	}

	return &AnnotatePatternNode{
		Pattern: left,
		Type:    right,
		Facts:   database.NewFacts(span()),
	}, nil
}

func (node *AnnotatePatternNode) Visit(visitor *visit.Visitor) {
	database.HideNode[*AnnotatePatternNode]()

	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node

	visitor.Visit(node.Pattern)
	visitor.Visit(node.Type)

	visitor.Constraint(typecheck.NewGroupConstraint(node.Pattern, node.Type))
	visitor.Constraint(typecheck.NewGroupConstraint(node, node.Pattern))
}

func (node *AnnotatePatternNode) Codegen(c *codegen.Codegen) error {
	return c.Write(node.Pattern)
}

func (node *AnnotatePatternNode) EachTemporary(f func(database.Node)) {
	EachTemporary(node.Pattern, f)
}

File: compiler/src/nodes/patterns/annotate.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::{InheritTemporaries, parse_pattern_element, parse_type_element, visit_pattern},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::GroupConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct AnnotatePatternNode {
    pub pattern: NodeRef,
    pub ty: NodeRef,
}

impl Node for AnnotatePatternNode {}

pub fn parse_annotate_pattern(parser: &mut Parser<'_>) -> Result<AnnotatePatternNode, ParseError> {
    let pattern = parse_pattern_element(parser)?;
    parser.token(TokenKind::AnnotateOperator)?;
    parser.commit("in this type annotation");
    parser.consume_line_breaks();
    let ty = parse_type_element(parser)?;

    Ok(AnnotatePatternNode { pattern, ty })
}

impl Visit for AnnotatePatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_pattern(node, visitor);

        visitor.visit(&self.pattern);
        visitor.edge(&self.pattern, node, "pattern");

        visitor.visit(&self.ty);
        visitor.edge(&self.ty, node, "type");

        visitor.constraint(GroupConstraint::new(self.pattern.clone(), self.ty.clone()));
        visitor.constraint(GroupConstraint::new(node.clone(), self.pattern.clone()));

        visitor.insert(node, InheritTemporaries(vec![self.pattern.clone()]));
    }
}

impl Codegen for AnnotatePatternNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        ctx.write(&self.pattern)?;
        Ok(())
    }
}
